mapping_parenting_tech/analysis/prototyping/play_store/get_play_store_apps.py

```diff
@@ -7,141 +7,413 @@
+    """
+    Saves JSON representation of app details. Will append to file <filename> if it already exists.
+    Returns True if exits successfully
+    """
+
+    output_target = PROJECT_DIR / "outputs/data/" / filename
+    with open(output_target, "at") as output_file:  # append in text mode
+        json.dump(details_dict, output_file, indent=2, default=str)
+
+    return True
+
 
 # %%
-output_target = OUTPUT_PATH / OUTPUT_PLAY_DETAILS_FILE
-with open(output_target, "w") as output_file:
-    json.dump(all_app_details, output_file, indent=2, default=str)
-output_file.close
+def load_app_details(filename: str) -> json:
+    """
+    Loads JSON file containing app details from <filename>. Returns JSON object.
+    """
+
+    result = {}
+    with open(PROJECT_DIR / filename, "rt") as input_file:
+        result = json.load(input_file)
+
+    return result
 
-print(f"{len(all_app_details)} apps' details saved to {output_target}")
 
 # %% [markdown]
-# Download reviews for the apps in the list
+# ## Download reviews for the apps in a given list of app ids
+# These functions enable iterative downloading into a single file, which is appended as more reviews are added
+#
+# `get_playstore_app_reviews` gets up to 200 reviews for a single app and returns them as a list of dicts (JSON), while `get_playstore_app_list_reviews` retrievs reviews for a given list of apps and saves them to a CSV file.
 
 # %%
-all_app_reviews = dict()
-running_total = 0
-
-for app_id in tqdm(output_ids["Parenting apps"], desc="Retrieving app reviews"):
-    try:
-        app_reviews = reviews_all(
-            app_id,
-            sleep_milliseconds=0,  # defaults to 0
+def get_playstore_app_reviews(
+    target_app_id, how_many: int = 200, continuation_token: str = None
+) -> [[{}], object]:
+    """
+    Returns up to 200 reviews for a given app in the Google Play Store. If more reviews are available, 'continuation_token' is also returned, which can then
+    be passed as a parameter to indicate where the function should resume.
+
+    Arguments:
+        target_app_id: id of the app for which you want to download reviews
+        how_many: number of reviews you wish to download in one grab; defaults to 200, which is the maximum option
+        continuation_token: indicates that the function should continue fetching reviews from this point
+
+    Returns
+        1. a list of reviews, each as a dict (see below)
+        2. continuation_token: object to pass back to this function to pick up where it left off
+
+    The reviews themselves are dicts in the following format:
+        "userName": str,
+        "userImage": str,
+        "content": str,
+        "score": int,
+        "thumbsUpCount": int,
+        "reviewCreatedVersion": str,
+        "at": datetime.datetime,
+        "replyContent": str,
+        "repliedAt": datetime.datetime,
+        "reviewId": str
+    """
+
+    if target_app_id == "":
+        return {"No app id given"}
```

Comment:
Does raising an error cause the function to return to where it was called? I'm wondering how this would be handled subsequently. I was trying to give the user the option of catching the 'error' quietly (although I realise that I didn't take advantage of this myself when I call this function!)

Reply:
Using `logging.warning` won't raise an error - it's just a type of a log message (as opposed to `logging.info`). You should still return something (eg `None`) so that it the function is terminated.